### PR TITLE
Add support for `pytest.approx` comparisons between array and scalar

### LINKED ISSUE #3312: pytest.approx doesn't support comparison between arrays and scalars
Sometimes we want to assert that all elements from an array are approximately equal to a single value, so this code comes naturally to mind:

```python
assert some_array == approx(1.0)
```

the problem is that `pytest.approx` doesn't support this (the workaround is to create a second array with the same shape and filled with the scalar).

Here is a test case:

```python
from pytest import approx
import numpy as np

def test_approx_array_scalar():
    assert np.array([1.0, 1.0]) == approx(1.0)

def test_approx_scalar_array():
    assert 1.0 == approx(np.array([1.0, 1.0]))
```

here are the outputs:

```python
______________________________________________ test_approx_array_scalar _______________________________________________
                                                                                                                       
    def test_approx_array_scalar():                                                                                    
>       assert np.array([1.0, 1.0]) == approx(1.0)                                                                     
E       ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()        
                                                                                                                       
test_approx.py:5: ValueError                                                                                           
______________________________________________ test_approx_scalar_array _______________________________________________
                                                                                                                       
    def test_approx_scalar_array():                                                                                    
>       assert 1.0 == approx(np.array([1.0, 1.0]))                                                                     
E       assert 1.0 == approx([1.0 ± 1.0e-06, 1.0 ± 1.0e-06])                                                           
E        +  where approx([1.0 ± 1.0e-06, 1.0 ± 1.0e-06]) = approx(array([ 1.,  1.]))                                   
E        +    where array([ 1.,  1.]) = <built-in function array>([1.0, 1.0])                                          
E        +      where <built-in function array> = np.array                                                             
                                                                                                                       
test_approx.py:8: AssertionError                                                                                       
```

This is `pip list`, including pytest version:
```
certifi (2016.2.28)
colorama (0.3.9)   
numpy (1.13.1)     
pip (9.0.1)        
py (1.4.34)        
pytest (3.2.1)     
setuptools (36.4.0)
wheel (0.29.0)     
wincertstore (0.2) 
```

### PR DESCRIPTION
**Pull request by tadeu**
fixes #3312

### CODE REVIEW

**Comment by tadeu** [comment]
Note that this only works in `numpy >=1.13`

**Comment by nicoddemus** [comment]
@kalekundert would you like to take a look as well?

File: _pytest/python_api.py [hunk 1]
```diff
@@ -100,8 +103,16 @@ def _yield_comparisons(self, actual):
         # We can be sure that `actual` is a numpy array, because it's
```
[the remark this answers is not in the document]
Comment (tadeu) [reply _pytest/python_api.py]: Sure, I'm changing it to

```python
        # For both `actual` and `self.expected`, they can independently be
        # either a `numpy.array` or a scalar. They are treated in `__eq__`
        # before being passed to `ApproxBase.__eq__`, which is the only
        # method that calls this one.
```

File: _pytest/python_api.py [hunk 2]
```diff
@@ -69,9 +73,6 @@ class ApproxNumpy(ApproxBase):
     Perform approximate comparisons for numpy arrays.
     """
 
-    # Tell numpy to use our `__eq__` operator instead of its.
-    __array_priority__ = 100
-
     def __repr__(self):
```
[the remark this answers is not in the document]
Comment (tadeu) [reply _pytest/python_api.py]: You're right, I'll to improve this

**Comment by tadeu** [comment]
@kalekundert all done, want to take another look?

**Comment by kalekundert** [comment]
Either way, `ApproxScalar.__repr__` is going to be called to get the nice message to display.  That won't be affected by anything that happens in `ApproxScalar.__eq__`.  In other words, the code responsible for making a nice message (if the expected value is a scalar) is `ApproxScalar.__repr__`, and it doesn't even know about the actual value (or whether or not it's an array).  The `ApproxScalar.__eq__` method just needs to return true or false, correctly taking into account the tolerances.

The more I think about it, the more I think `ApproxNumpy.expected` should always be an array.  Getting philosophical again, the `Approx*` classes are basically wrappers around an expected value, and it doesn't really make sense for there to be two ways to wrap a scalar (especially since one would only work if you then subsequently compared it to an array).

Anyways, I'm worried that I'm losing sight of the fact that the code is totally fine.  :-)  So I'll leave it to your judgment.  If you think it'd be better (and worth the effort) to require `ApproxNumpy.expected` to be an array, go ahead and do it.  Otherwise let me know and I'll merge.

**Comment by tadeu** [comment]
Hi @kalekundert, I agree with you, I've changed the code so that `self.expected` in `ApproxNumpy` is always an array now, the code is simpler ;)

**Comment by kalekundert** [comment]
Awesome, thanks for the PR!

**Comment by nicoddemus** [comment]
Great, thanks @tadeu and @kalekundert!

**Pull request merged**